kernel/kernel-api/src/main/java/io/delta/kernel/types/StructField.java

```diff
@@ -48,31 +50,96 @@ public class StructField {
           METADATA_ROW_INDEX_COLUMN_NAME,
           LongType.LONG,
           false,
-          FieldMetadata.builder().putBoolean(IS_METADATA_COLUMN_KEY, true).build());
+          FieldMetadata.builder().putBoolean(IS_METADATA_COLUMN_KEY, true).build(),
+          Collections.emptyList());
 
   public static final String COLLATIONS_METADATA_KEY = "__COLLATIONS";
 
   ////////////////////////////////////////////////////////////////////////////////
   // Instance Fields / Methods
   ////////////////////////////////////////////////////////////////////////////////
 
+  /**
+   * Represents a type change for a field, containing the original and new primitive types.
+   *
+   * <p>Type changes are actually persisted in metadata attached to StructFields but the rules for
+   * where the metadata is attached depend on if the change is for nested arrays/maps or primitive
+   * types.
+   */
+  public static class TypeChange {
+    private final BasePrimitiveType from;
+    private final BasePrimitiveType to;
+
+    public TypeChange(BasePrimitiveType from, BasePrimitiveType to) {
+      this.from = Objects.requireNonNull(from, "from type cannot be null");
+      this.to = Objects.requireNonNull(to, "to type cannot be null");
+    }
+
+    public BasePrimitiveType getFrom() {
+      return from;
+    }
+
+    public BasePrimitiveType getTo() {
+      return to;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      TypeChange that = (TypeChange) o;
+      return Objects.equals(from, that.from) && Objects.equals(to, that.to);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(from, to);
+    }
+
+    @Override
+    public String toString() {
+      return String.format("TypeChange(from=%s,to=%s)", from, to);
+    }
+  }
+
   private final String name;
   private final DataType dataType;
   private final boolean nullable;
   private final FieldMetadata metadata;
+  private final List<TypeChange> typeChanges;
 
   public StructField(String name, DataType dataType, boolean nullable) {
     this(name, dataType, nullable, FieldMetadata.empty());
   }
 
   public StructField(String name, DataType dataType, boolean nullable, FieldMetadata metadata) {
+    this(name, dataType, nullable, metadata, Collections.emptyList());
+  }
+
+  public StructField(
+      String name,
+      DataType dataType,
+      boolean nullable,
+      FieldMetadata metadata,
+      List<TypeChange> typeChanges) {
     this.name = name;
     this.dataType = dataType;
     this.nullable = nullable;
+    this.typeChanges = typeChanges == null ? Collections.emptyList() : typeChanges;
 
     FieldMetadata collationMetadata = fetchCollationMetadata();
     this.metadata =
         new FieldMetadata.Builder().fromMetadata(metadata).fromMetadata(collationMetadata).build();
+    if (!this.typeChanges.isEmpty()
```

Comment:
if the map contains a struct either as key or value, is the type change allowed in this case?

Reply:
Yes, the idea of this assert is it enforces types changes are only recorded on leaf nodes of the schema.  We could also assert this in the other way I think that DataType is a primitive type.

I think this is a core design question that we should answer which option we want to pursue here. 
1.  By adding these to leaf nodes in the schema we have to add code similar to collations that will construct appropriate FieldMetadata on construction (and do the inverse operation for when parsing) for nested Map/Array types.
2. The other option would be not add the field at all and solely work with TypeChanges are a function of FieldMetadata (which would require less book-keeping but might be more error prone in the long run.  This is how  field assignment works for icebergCompatV2

Reply:
I think the question is which pattern we want to follow (and then migrate all of these common elements to the common pattern).

Reply:
@allisonport-db this is somewhat related to the bug on populating nested field IDs, not sure if you have thoughts on which approach you prefer.

Reply:
I am a little confused because I think this is the level where we would want the type changes for map/array? From the protocol https://github.com/delta-io/delta/blob/master/PROTOCOL.md#type-change-metadata

Reply:
We wouldn't want to support TypeChanges on their element or key/map types though.

I think I maybe understand what you're saying however, we could choose to _store_ them at either level and then serialize them at this level?

I honestly find the use of StructField for the key/value/element really confusing since there is no such thing as a key's fieldMetadata (it does not exist!). For that reason, I'd prefer to store it here at this level. And honestly I've been meaning to take a look at that code and see if there's a better way to make sure there isn't confusing when using StructField in those data types.

Reply:
But then we need to add the additional "fieldPath" to the TypeChange class right?

Reply:
Also feel free to disagree, I think choosing where to store this info (and other stuff like the nested ids) is an important discussion to be had